test/test_core.py

```diff
@@ -7804,6 +7803,34 @@ def test_embind_val_cross_thread(self):
     ''')
     self.do_runf('test_embind_val_cross_thread.cpp', 'val accessed from wrong thread', assert_returncode=NON_ZERO)
 
+  @node_pthreads
+  def test_embind_val_cross_thread_deleted(self):
+    self.emcc_args += ['--bind']
+    create_file('test_embind_val_cross_thread.cpp', r'''
+      #include <emscripten.h>
+      #include <emscripten/val.h>
+      #include <thread>
+      #include <stdio.h>
+      #include <optional>
+
+      using emscripten::val;
+
+      int main(int argc, char **argv) {
+        // Create a storage for value handles on the main thread.
+        std::optional<val> opt_value;
+        std::thread([&] {
+          // Set to a value handle from a different thread.
+          val& value = opt_value.emplace(1);
+          // Move out from the optional storage so that we free the value on the same thread.
+          val moved_out = std::move(value);
+        }).join();
+        // Now std::optional is initialized but with a deleted value handle.
+        // There should be no cross-thread error here when it tries to free that value,
+        // because the value has already been deleted on the correct thread.
+      }
+    ''')
+    self.do_runf('test_embind_val_cross_thread.cpp')
```

Comment:
I wonder if both of these tests should be in test_other instead.. then they don't get run in so many configurations and use less CI time?

Reply:
Hm most Embind tests live in core, I assumed it's intentional? But I don't know the actual rules and usually just follow the precedent.

Reply:
TIL. In that case, yeah, quite a lot of Embind tests could live in other.

Reply:
OK with me to leave that as a followup.